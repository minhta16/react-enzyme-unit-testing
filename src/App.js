import React from "react";
import logo from "./logo.svg";
import "./App.css";

const Test = () => <div>Testing!</div>;
class App extends React.Component {
  state = {
    on: false,
    text: ""
  };

  handleStrings = string => {
    return string === "Hello World";
  };
  componentDidMount() {}
  componentWillReceiveProps() {}
  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <p>
            Edit <code>src/App.js</code> and save to reload.
          </p>
          <ul className="yolo">
            <li>1</li>
            <li>2</li>
            <li>3</li>
          </ul>
          <a
            className="App-link"
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
          >
            Learn React
          </a>
          <button onClick={e => this.setState({ on: true })}>Click</button>
          <div className="button-state">{this.state.on ? "Yes!" : "No :("}</div>
          <input
            onChange={e => this.setState({ text: e.currentTarget.value })}
          />
          <div className="input-text">{this.state.text}</div>
        </header>
        <Test />
      </div>
    );
  }
}

export class Link extends React.Component {
  render() {
    return this.props.hide ? null : <a href={this.props.address}>Click!</a>;
  }
}
export default App;
