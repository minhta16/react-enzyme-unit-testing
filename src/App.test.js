import React from "react";
import ReactDOM from "react-dom";
import App, { Link } from "./App";
import { configure, shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

configure({ adapter: new Adapter() });

describe("<App /> shallow rendering", () => {
  it("should render App", () => {
    const wrapper = shallow(<App />);
    console.log(wrapper);
  });

  it("should contain 1 p elements", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("p").length).toBe(1);
    console.log(wrapper);
  });

  it("should contain className App-header", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(".App-header").exists()).toBe(true);
    console.log(wrapper);
  });

  it("should contain 3 ul-li elements", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("ul").children().length).toBe(3);
    console.log(wrapper);
  });

  it("should contain ul with className yolo", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("ul").hasClass("yolo")).toBe(true);
    console.log(wrapper);
  });

  it("matches the snapshot", () => {
    const tree = shallow(<App />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it("on button click changes div text", () => {
    const wrapper = shallow(<App />);
    const button = wrapper.find("button");
    expect(wrapper.find(".button-state").text()).toBe("No :(");
    button.simulate("click");
    expect(wrapper.find(".button-state").text()).toBe("Yes!");
  });

  it("on input change changes text", () => {
    const wrapper = shallow(<App />);
    const input = wrapper.find("input");
    expect(wrapper.find(".input-text").text()).toBe("");
    input.simulate("change", { currentTarget: { value: "Minh" } });
    expect(wrapper.find(".input-text").text()).toBe("Minh");
  });

  // Test with life cycle methods
  it("calls componentDidMount", () => {
    jest.spyOn(App.prototype, "componentDidMount");
    const wrapper = shallow(<App />);
    expect(App.prototype.componentDidMount.mock.calls.length).toBe(1);
  });

  it("setProps calls componentWillReceiveProps", () => {
    jest.spyOn(App.prototype, "componentWillReceiveProps");
    const wrapper = shallow(<App />);
    wrapper.setProps({ hide: true });
    expect(App.prototype.componentWillReceiveProps.mock.calls.length).toBe(1);
  });

  // Test with methods
  it("handleStrings function returns correctly", () => {
    const wrapper = shallow(<App />);
    const trueReturn = wrapper.instance().handleStrings("Hello World");
    expect(trueReturn).toBe(true);
  });
});

describe("<App /> mount rendering", () => {
  it("should render App", () => {
    const wrapper = mount(<App />);
    console.log(wrapper);
    wrapper.unmount();
  });

  it("matches the snapshot", () => {
    const tree = mount(<App />);
    expect(toJson(tree)).toMatchSnapshot();
    tree.unmount();
  });
});

describe("<Link />", () => {
  it("link component accept addres props", () => {
    const wrapper = shallow(<Link address="www.google.com" />);
    expect(wrapper.instance().props.address).toBe("www.google.com");
  });

  it("a tag node renders href correctly", () => {
    const wrapper = shallow(<Link address="www.google.com" />);
    expect(wrapper.props().href).toBe("www.google.com");
  });

  it("returns null with true hide prop", () => {
    const wrapper = shallow(<Link hide={false} />);
    expect(wrapper.find("a").length).toBe(1);
    wrapper.setProps({ hide: true });
    expect(wrapper.get(0)).toBeNull();
  });
});
