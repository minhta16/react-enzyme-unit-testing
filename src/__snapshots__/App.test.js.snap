// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> mount rendering matches the snapshot 1`] = `
<App>
  <div
    className="App"
  >
    <header
      className="App-header"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
      <p>
        Edit 
        <code>
          src/App.js
        </code>
         and save to reload.
      </p>
      <ul
        className="yolo"
      >
        <li>
          1
        </li>
        <li>
          2
        </li>
        <li>
          3
        </li>
      </ul>
      <a
        className="App-link"
        href="https://reactjs.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn React
      </a>
      <button
        onClick={[Function]}
      >
        Click
      </button>
      <div
        className="button-state"
      >
        No :(
      </div>
      <input
        onChange={[Function]}
      />
      <div
        className="input-text"
      />
    </header>
    <Test>
      <div>
        Testing!
      </div>
    </Test>
  </div>
</App>
`;

exports[`<App /> shallow rendering matches the snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <p>
      Edit 
      <code>
        src/App.js
      </code>
       and save to reload.
    </p>
    <ul
      className="yolo"
    >
      <li>
        1
      </li>
      <li>
        2
      </li>
      <li>
        3
      </li>
    </ul>
    <a
      className="App-link"
      href="https://reactjs.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Learn React
    </a>
    <button
      onClick={[Function]}
    >
      Click
    </button>
    <div
      className="button-state"
    >
      No :(
    </div>
    <input
      onChange={[Function]}
    />
    <div
      className="input-text"
    />
  </header>
  <Test />
</div>
`;
